app/demo-loop/LDemoIO.root.cc

```diff
@@ -0,0 +1,67 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2022 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file demo-loop/LDemoIO.root.cc
+//---------------------------------------------------------------------------//
+#include "LDemoIO.hh"
+
+#include <TBranch.h>
+#include <TFile.h>
+#include <TTree.h>
+
+namespace demo_loop
+{
+//---------------------------------------------------------------------------//
+void to_root(std::shared_ptr<celeritas::RootFileManager> root_manager,
+             LDemoArgs&                                  args)
+{
+    CELER_EXPECT(root_manager);
+    CELER_EXPECT(args);
+
+    // So far RootFileManager has only a single TFile instance and ROOT knows
+    // that this new TTree has to be attached to this single TFile. If we
+    // expand it to a RootFileManager::tfile_[tid] we'd need to specify the
+    // correct TFile* by invoking TTree("name", "title", TFile*).
+    std::unique_ptr<TTree> tree_input
+        = std::make_unique<TTree>("input", "input");
+
+    // Problem definition
+    tree_input->Branch("geometry_filename", &args.geometry_filename);
+    tree_input->Branch("physics_filename", &args.physics_filename);
+    tree_input->Branch("hepmc3_filename", &args.hepmc3_filename);
+
+    // Control
+    tree_input->Branch("seed", &args.seed);
+    // {
+    // Celeritas' type aliases have to be manually casted to be stored properly
+    // See leaflist types: https://root.cern.ch/doc/master/classTBranch.html
+    tree_input->Branch(
+        "max_num_tracks", &args.max_num_tracks, "max_num_tracks/l");
+    tree_input->Branch("max_steps", &args.max_steps, "max_num_steps/l");
+
+    tree_input->Branch("initializer_capacity",
+                       &args.initializer_capacity,
+                       "initializer_capacity/l");
+    tree_input->Branch("max_events", &args.max_events, "max_events/l");
+    tree_input->Branch("secondary_stack_factor",
+                       &args.secondary_stack_factor,
+                       "secondary_stack_factor/D");
+    // }
+    tree_input->Branch("enable_diagnostics", &args.enable_diagnostics);
+    tree_input->Branch("use_device", &args.use_device);
+    tree_input->Branch("sync", &args.sync);
+    tree_input->Branch("step_limiter", &args.step_limiter);
+
+    // Options for physics processes and models
+    tree_input->Branch("combined_brem", &args.brem_combined);
+
+    // TODO Add magnetic field information?
+
+    // Fill tree and write it to file
+    tree_input->Fill();
+    tree_input->Write();
```

Comment:
I am confused.  Because of the unique pointer this routine create a `TTree` with a single entry !?  Is that the intent?

Reply:
a `TTree` is a lot of over-head for a single entry of a small object.  Is there advantages over just storing the `args` object itself/directly?

Reply:
What do you mean by storing the object directly? You mean doing something like this?
```cpp
TTree tree_input("input", "input");
tree_input.Branch("input_args", &args);
tree_input.Fill();
tree_input.Write();
```

I went for a bunch of basic branches so we don't need anything other than root to read the output files.

Reply:
I meant:
```
file->WriteObject(&args, "SimInputArgs");
```
and indeed, it then requires the 'dictionary' for the type of `args` but the size on file would be much less than going through a `TTree` (whether the space saving is worth the 'cost' of generating the dictionary depends on a bit on how you plan to use/read the data).

Reply:
It would mean adding more ROOT stuff to the app framework, which I don't think is worth it. Maybe it's better just to not have the ROOT copy of the input args at all?

Reply:
We could, but then we might mix up multiple output data while testing/validating the code, no matter how well we name the root files. For context the input TTree costs a few kB. Any tiny simulation will result in MB sized files anyway, so I think there is no real downside in storing the input, even without being a smaller object. And as we said, anyone can, out of the box, fully read the data without the hassle of loading a dictionary.

Reply:
Yeah fine, I'm cool with the few-kB overhead of a tree, but not so much with adding more ROOT link dictionary complications to the demo app. We should probably have our python inputs generate unique IDs as well as correlating the input/output/root filenames.

Reply:
>  fully read the data without the hassle of loading a dictionary.

Just a reminder that ROOT files are self describing so you can always read the content even without the dictionary (*), however it might require a step or two more compare to just storing the data in a `TTree`.

> For context the input TTree costs a few kB. Any tiny simulation will result in MB sized files anyway, 

Fair enough.

(*) except for the case where the user explicitly implemented a custom streamer.